presto-router/src/test/java/com/facebook/presto/router/TestSelectors.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.router;
+
+import com.facebook.airlift.bootstrap.Bootstrap;
+import com.facebook.airlift.bootstrap.LifeCycleManager;
+import com.facebook.airlift.http.server.testing.TestingHttpServerModule;
+import com.facebook.airlift.jaxrs.JaxrsModule;
+import com.facebook.airlift.json.JsonModule;
+import com.facebook.airlift.log.Logging;
+import com.facebook.airlift.node.testing.TestingNodeModule;
+import com.facebook.presto.router.cluster.ClusterManager;
+import com.facebook.presto.router.cluster.RequestInfo;
+import com.facebook.presto.server.security.ServerSecurityModule;
+import com.facebook.presto.server.testing.TestingPrestoServer;
+import com.facebook.presto.testing.assertions.Assert;
+import com.facebook.presto.tpch.TpchPlugin;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Injector;
+import com.google.inject.Module;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CLIENT_TAGS;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_SOURCE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_USER;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+
+public class TestSelectors
+{
+    private List<TestingPrestoServer> prestoServers;
+    private ClusterManager clusterManager;
+    private LifeCycleManager lifeCycleManager;
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        String configTemplate = new String(
+                Files.readAllBytes(Paths.get(this.getClass().getClassLoader().getResource("selector-test-router-template.json").getPath())));
+
+        File configFile = File.createTempFile("router", "json");
+
+        Logging.initialize();
```

Comment:
Instead of this setup, I recommend just passing a config object directly to the `RouterModule` with a new constructor that accepts the config object as an argument